File: bot/handlers/admin_callbacks.py
from aiogram import Router, F
from aiogram.types import CallbackQuery
from bot.config import config
from bot.keyboards.admin_kb import get_admin_keyboard, get_admin_sub_keyboard
from bot.utils.db import (
    get_stats,
    get_recent_users,
    get_recent_messages,
    get_visitor_stats,
    get_recent_visitors
)

router = Router()

def is_admin(user_id: int) -> bool:
    return user_id in config.ADMIN_IDS or str(user_id) in [str(a) for a in config.ADMIN_IDS]

def build_admin_dashboard_text(user_id: int) -> str:
    stats = get_stats()
    v_stats = get_visitor_stats()
    return f"""👑 <b>BEKZOD IDIYEV — ADMIN BOSHQARUV MARKAZI</b>
━━━━━━━━━━━━━━━━━━━━━━━━━━
🆔 <b>Admin ID:</b> <code>{user_id}</code>
⚡ <b>Engine:</b> 🟢 Python aiogram 3.x + Serverless Gateway
🌐 <b>Portfolio:</b> <code>{config.PORTFOLIO_URL}</code>
🐙 <b>GitHub:</b> <code>{config.GITHUB_URL}</code>

📈 <b>JONLI STATISTIKA:</b>
• 🌐 <b>Portfolio Tashriflari:</b> <code>{v_stats['total_visits']}</code> ta (Bugun: <code>{v_stats['today_visits']}</code>)
• 👥 <b>Bot Foydalanuvchilari:</b> <code>{stats['total_users']}</code> ta
• 📩 <b>Qabul Qilingan Xabarlar:</b> <code>{stats['total_messages']}</code> ta

Quyidagi bo'limlardan birini tanlang:"""

@router.callback_query(F.data == "admin_main")
async def cb_admin_main(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    dashboard_text = build_admin_dashboard_text(callback.from_user.id)
    await callback.message.edit_text(dashboard_text, reply_markup=get_admin_keyboard(), parse_mode="HTML")
    await callback.answer()

@router.callback_query(F.data == "admin_stats")
async def cb_admin_stats(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    stats = get_stats()
    lang_info = "".join([f"• {'🇺🇿' if l=='uz' else ('🇷🇺' if l=='ru' else '🇬🇧')} <b>{l.upper()}:</b> {c} ta\n" for l, c in stats.get("languages", {}).items()]) or "  • <i>Til ma'lumotlari to'planmoqda</i>\n"
    stats_text = f"""📊 <b>TO'LIQ BOT STATISTIKASI VA METRIKALAR</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n👥 <b>Jami Bot A'zolari:</b> <code>{stats['total_users']}</code> ta\n📩 <b>Qabul Qilingan Xabarlar:</b> <code>{stats['total_messages']}</code> ta\n\n🌐 <b>Til Bo'yicha:</b>\n{lang_info}\n⚡ <b>Tizim Holati:</b>\n• Database: SQLite WAL Mode\n• Routing: aiogram 3.x"""
    await callback.message.edit_text(stats_text, reply_markup=get_admin_sub_keyboard("admin_stats"), parse_mode="HTML")
    await callback.answer()

@router.callback_query(F.data == "admin_visitors")
async def cb_admin_visitors(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    v_stats = get_visitor_stats()
    cities = "".join([f"  • 🏙️ <b>{c}:</b> <code>{cnt}</code> ta\n" for c, cnt in v_stats.get("top_cities", [])]) or "  • <i>Hozircha ma'lumotlar to'planmoqda</i>\n"
    devices = "".join([f"  • {d or '💻 Desktop'}: <code>{cnt}</code> ta\n" for d, cnt in v_stats.get("devices", {}).items()]) or "  • <i>Qurilma ma'lumotlari yo'q</i>\n"
    visitors_text = f"""🌐 <b>PORTFOLIO SAYTI TASHRIFLAR STATISTIKASI</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n👥 <b>Jami Tashriflar:</b> <code>{v_stats['total_visits']}</code> ta\n📅 <b>Bugungi Tashriflar:</b> <code>{v_stats['today_visits']}</code> ta\n🔑 <b>Noyob Mehmonlar (IP):</b> <code>{v_stats['unique_ips']}</code> ta\n\n📱 <b>Qurilmalar Bo'yicha:</b>\n{devices}\n🌍 <b>Top Shaharlar:</b>\n{cities}\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n⚡ <i>Real-vaqt monitoringi faol</i>"""
    await callback.message.edit_text(visitors_text, reply_markup=get_admin_sub_keyboard("admin_visitors"), parse_mode="HTML")
    await callback.answer()

@router.callback_query(F.data == "admin_recent_visitors")
async def cb_admin_recent_visitors(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    visitors = get_recent_visitors(5)
    if not visitors:
        text = "👁️ <b>SO'NGGI MEHMONLAR</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n<i>Hozircha yangi mehmonlar bazada mavjud emas.</i>"
    else:
        v_lines = ""
        for i, v in enumerate(visitors, 1):
            name = v["visitor_name"] or "Anonim"
            city = v["city"] or "O'zbekiston"
            dev = v["os"] or v["device_type"] or "Desktop"
            map_link = f" | <a href='https://yandex.uz/maps/?pt={v['longitude']},{v['latitude']},pm2rdm&z=16'>📍 Xarita</a>" if v['latitude'] and v['longitude'] else ""
            v_lines += f"<b>{i}. {name}</b>\n📍 {city} ({dev}){map_link}\n🕒 <i>{v['visited_at']}</i>\n\n"
        text = f"👁️ <b>SO'NGGI 5 TA PORTFOLIO MEHMONI:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n{v_lines}"
    await callback.message.edit_text(text, reply_markup=get_admin_sub_keyboard("admin_recent_visitors"), parse_mode="HTML")
    await callback.answer()

@router.callback_query(F.data == "admin_users")
async def cb_admin_users(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    users = get_recent_users(5)
    user_lines = "".join([f"<b>{i}. {u['full_name']}</b> (@{u['username'] or 'mavjud_emas'})\n🆔 <code>{u['user_id']}</code> | 🌐 {u['language'].upper()}\n🕒 <i>{u['last_active']}</i>\n\n" for i, u in enumerate(users, 1)])
    users_text = f"👥 <b>SO'NGGI 5 TA BOT FOYDALANUVCHISI:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n{user_lines or '<i>Foydalanuvchilar yoq.</i>'}"
    await callback.message.edit_text(users_text, reply_markup=get_admin_sub_keyboard("admin_users"), parse_mode="HTML")
    await callback.answer()

@router.callback_query(F.data == "admin_leads")
async def cb_admin_leads(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    messages = get_recent_messages(5)
    msg_lines = "".join([f"<b>{i}. {m['user_name']}</b> (<code>{m['contact_info']}</code>)\n💬 \"<i>{m['message_text'][:100]}</i>\"\n🕒 <i>{m['created_at']}</i>\n\n" for i, m in enumerate(messages, 1)])
    leads_text = f"📩 <b>SO'NGGI 5 TA QABUL QILINGAN XABAR:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n{msg_lines or '<i>Xabarlar mavjud emas.</i>'}"
    await callback.message.edit_text(leads_text, reply_markup=get_admin_sub_keyboard("admin_leads"), parse_mode="HTML")
    await callback.answer()

@router.callback_query(F.data == "admin_diag")
async def cb_admin_diag(callback: CallbackQuery):
    if not callback.from_user or not is_admin(callback.from_user.id):
        await callback.answer("Ruxsat berilmagan.", show_alert=True)
        return
    diag_text = f"""⚡ <b>TIZIM DIAGNOSTIKASI VA SALOMATLIK HOLATI</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n✅ <b>Python aiogram Engine:</b> 3.x Active\n✅ <b>Lokal Ma'lumotlar Bazasi:</b> SQLite Active\n✅ <b>Vercel Serverless Webhook:</b> 24/7 Connected\n✅ <b>Portfolio URL:</b> <code>{config.PORTFOLIO_URL}</code>\n✅ <b>Admin Autentifikatsiyasi:</b> <code>{callback.from_user.id}</code> (Tasdiqlangan)\n✅ <b>Anti-Spam:</b> Faol"""
    await callback.message.edit_text(diag_text, reply_markup=get_admin_sub_keyboard("admin_diag"), parse_mode="HTML")
    await callback.answer()

File: api/bot/handlers/adminCallbacks.ts
import { escapeHtml, buildAdminMainText } from '../telegramApi';
import {
  getVisitorStatsDb,
  getRecentVisitorsDb,
  getBotUserStatsDb,
  getGeoAndReferrerStatsDb,
  getRecentBotUsersDb,
  getRecentMessagesDb,
} from '../db';
import {
  getAdminMainKeyboard,
  getAdminSubKeyboard,
  getProjectDetailKeyboard,
  getProjectListKeyboard,
  PORTFOLIO_URL,
} from '../keyboards';

export async function handleAdminAndProjectCallbacks(
  data: string,
  serverTimestamp: string,
  chatId: string | number,
  botToken: string,
  botUserCount: number,
  totalInteractions: number
): Promise<{ text: string; markup: any } | null> {
  // Project detail callbacks
  if (data === 'proj_buddy') {
    return {
      text: `🚀 <b>BUDDY TEAM — AI MENTOR & TEAM MATCHMAKING</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n🎯 <b>Tavsif:</b> Talabalar va junior dasturchilarni qiziqishlari, texnologik darajasi va maqsadlariga qarab avtomatlashtirilgan AI algoritmi orqali jamoalarga birlashtiruvchi platforma.\n🛠️ <b>Stack:</b> Python 3.11, FastAPI, PostgreSQL, Redis, Docker, Celery\n⚡ <b>Natijalar:</b>\n• 500+ faol foydalanuvchilar o'rtasida muvaffaqiyatli match\n• 99.8% API uptime va asinxron fon vazifalari boshqaruvi`,
      markup: getProjectDetailKeyboard('proj_buddy'),
    };
  }
  if (data === 'proj_esports') {
    return {
      text: `🎮 <b>ESPORTS TOURNAMENT BOT</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n🎯 <b>Tavsif:</b> Kiberfutbol va kiberxavfsizlik turnirlarini avtomatlashtirilgan tarzda tashkil qilish, qoidalarni monitoring qilish va statistikani hisoblovchi yuqori yuklamali Telegram bot.\n🛠️ <b>Stack:</b> Python, aiogram 3.x, SQLite, Webhooks, Aiohttp\n⚡ <b>Afzalliklari:</b>\n• Avtomatlashtirilgan match grid va reyting hisoblagich\n• To'lov tizimlari va havola autentifikatsiyasi`,
      markup: getProjectDetailKeyboard('proj_esports'),
    };
  }
  if (data === 'proj_peerlearn') {
    return {
      text: `📚 <b>PEERLEARN TELEGRAM MINI APP</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n🎯 <b>Tavsif:</b> Telegram ichida to'liq ishlovchi interaktiv ta'lim Web App ekotizimi.\n🛠️ <b>Stack:</b> React, TypeScript, FastAPI, WebSockets\n⚡ <b>Imkoniyatlar:</b>\n• Darslar va testlarni Telegramdan chiqmasdan yechish\n• Jonli reyting va yutuqlar tizimi`,
      markup: getProjectDetailKeyboard('proj_peerlearn'),
    };
  }
  if (data === 'proj_list') {
    return {
      text: `📂 <b>ISHLAB CHIQARISH LOYIHALARI:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\nQuyidagi loyihalardan birini tanlang:\n\n1️⃣ <b>Buddy Team</b> — AI Mentor & Team Matchmaking\n2️⃣ <b>Esports Tournament Bot</b> — Turnir boshqaruv tizimi\n3️⃣ <b>PeerLearn Telegram Mini App</b> — Ta'lim platformasi`,
      markup: getProjectListKeyboard(),
    };
  }

  // Admin Dashboard callbacks
  if (data === 'admin_main') {
    const text = await buildAdminMainText(serverTimestamp, chatId, botUserCount);
    return { text, markup: getAdminMainKeyboard() };
  }
  if (data === 'admin_visitors') {
    const vStats = await getVisitorStatsDb();
    const total = vStats.totalVisits;
    const mobilePct = total > 0 ? Math.round((vStats.deviceStats.mobile / total) * 100) : 0;
    const desktopPct = total > 0 ? 100 - mobilePct : 0;
    let citiesStr = vStats.topCities.map((c) => `  • 🏙️ <b>${escapeHtml(c.city)}:</b> <code>${c.count}</code> ta`).join('\n');
    if (!citiesStr) citiesStr = "  • <i>Hozircha ma'lumotlar to'planmoqda</i>";

    const text = `🌐 <b>PORTFOLIO SAYTI REAL TASHRIFLAR STATISTIKASI</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n👥 <b>Jami Tashriflar:</b> <code>${vStats.totalVisits}</code> ta\n📅 <b>Bugungi Tashriflar:</b> <code>${vStats.todayVisits}</code> ta\n🔑 <b>Noyob Mehmonlar (IP):</b> <code>${vStats.uniqueIps}</code> ta\n\n📱 <b>Qurilmalar Bo'yicha:</b>\n  • 📱 Mobile: <code>${mobilePct}%</code>\n  • 💻 Desktop: <code>${desktopPct}%</code>\n\n🌍 <b>Top Shaharlar:</b>\n${citiesStr}\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n⚡ <i>PostgreSQL real-vaqt monitoringi faol</i>`;
    return { text, markup: getAdminSubKeyboard('admin_visitors') };
  }
  if (data === 'admin_recent_visitors') {
    const visitors = await getRecentVisitorsDb(5);
    let vLines = visitors.map((v: any, i: number) => {
      const name = v.visitor_name || 'Anonim';
      const city = v.city || v.country || "O'zbekiston";
      const dev = v.device_type || v.os || 'Desktop';
      const mapLink = v.latitude && v.longitude ? ` | <a href="https://yandex.uz/maps/?pt=${v.longitude},${v.latitude},pm2rdm&z=16">📍 Xarita</a>` : '';
      return `<b>${i + 1}. ${escapeHtml(name)}</b>\n📍 ${escapeHtml(city)} (${escapeHtml(dev)})${mapLink}\n🕒 <i>${v.visited_at || ''}</i>`;
    }).join('\n\n');
    const text = `👁️ <b>SO'NGGI ${visitors.length} TA REAL PORTFOLIO MEHMONI:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n${vLines || '<i>Hozircha yangi tashriflar yo\'q.</i>'}\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n⚡ <i>Avtomatik bazaga yoziladi.</i>`;
    return { text, markup: getAdminSubKeyboard('admin_recent_visitors') };
  }
  if (data === 'admin_stats') {
    const bStats = await getBotUserStatsDb();
    const vStats = await getVisitorStatsDb();
    const totalUsers = bStats.totalUsers > 0 ? bStats.totalUsers : botUserCount;
    const text = `📊 <b>TO'LIQ TELEMETRIYA VA REAL STATISTIKA</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n🤖 <b>Real Bot Foydalanuvchilari:</b> <code>${totalUsers}</code> ta\n🌐 <b>Portfolio Jami Tashriflari:</b> <code>${vStats.totalVisits}</code> ta\n⚡ <b>Jami Bajarilgan So'rovlar:</b> <code>${totalInteractions}</code> ta\n📩 <b>Yetkazilgan Xabarlar:</b> <code>${bStats.totalMessages}</code> ta\n\n⚡ <b>Infratuzilma:</b>\n• Database: Vercel PostgreSQL (Neon Serverless)\n• Hosting: Vercel Serverless Edge\n• Response Time: ~45ms\n• SSL: TLS 1.3 Active`;
    return { text, markup: getAdminSubKeyboard('admin_stats') };
  }
  if (data === 'admin_diag') {
    const pingStart = Date.now();
    let tgPing = 'N/A';
    try {
      await fetch(`https://api.telegram.org/bot${botToken}/getMe`);
      tgPing = `${Date.now() - pingStart}ms`;
    } catch {
      tgPing = 'Ulanishda xato';
    }
    const text = `⚡ <b>TIZIM DIAGNOSTIKASI VA SALOMATLIK HOLATI</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n✅ <b>Vercel Serverless Function:</b> 200 OK\n✅ <b>Telegram Webhook:</b> Ulangan & Faol\n✅ <b>Telegram Bot API Real Ping:</b> ${tgPing}\n✅ <b>Xavfsizlik & Anti-Spam:</b> Honeypot Trap & HTML Sanitizer Faol\n✅ <b>Admin Autentifikatsiyasi:</b> <code>${chatId}</code> (Tasdiqlangan)\n✅ <b>Database:</b> Neon PostgreSQL Connected`;
    return { text, markup: getAdminSubKeyboard('admin_diag') };
  }
  if (data === 'admin_geo') {
    const geoStats = await getGeoAndReferrerStatsDb();
    const countryLines = geoStats.countries.map((c) => {
      const pct = geoStats.total > 0 ? Math.round((c.count / geoStats.total) * 100) : 0;
      return `• 🌐 <b>${escapeHtml(c.country)}:</b> <code>${pct}%</code> (<code>${c.count}</code> ta)`;
    }).join('\n') || '• <i>Hozircha ma\'lumotlar mavjud emas</i>';
    const refLines = geoStats.referrers.map((r) => {
      const pct = geoStats.total > 0 ? Math.round((r.count / geoStats.total) * 100) : 0;
      return `• 🔗 <b>${escapeHtml(r.referrer)}:</b> <code>${pct}%</code> (<code>${r.count}</code> ta)`;
    }).join('\n') || '• <i>Hozircha ma\'lumotlar mavjud emas</i>';
    const text = `🌍 <b>TASHRIF BUYURUVCHILAR GEOGRAFIYASI:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n${countryLines}\n\n📊 <b>Manbalar (Referrers):</b>\n${refLines}\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n⚡ <i>Jami: ${geoStats.total} ta tashrif</i>`;
    return { text, markup: getAdminSubKeyboard('admin_geo') };
  }
  if (data === 'admin_users') {
    const users = await getRecentBotUsersDb(5);
    const uLines = users.map((u: any, i: number) => {
      const uname = u.username ? `@${escapeHtml(u.username)}` : 'mavjud_emas';
      return `<b>${i + 1}. ${escapeHtml(u.full_name)}</b> (${uname})\n🆔 <code>${u.user_id}</code> | 🌐 ${(u.language || 'uz').toUpperCase()}\n🕒 <i>${u.last_active}</i>`;
    }).join('\n\n');
    const text = `👥 <b>SO'NGGI ${users.length} TA BOT FOYDALANUVCHISI:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n${uLines || '<i>Hozircha foydalanuvchilar yo\'q.</i>'}\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n⚡ <i>Real PostgreSQL bazasi</i>`;
    return { text, markup: getAdminSubKeyboard('admin_users') };
  }
  if (data === 'admin_leads') {
    const messages = await getRecentMessagesDb(5);
    const mLines = messages.map((m: any, i: number) => {
      return `<b>${i + 1}. ${escapeHtml(m.user_name)}</b> (<code>${escapeHtml(m.contact_info)}</code>)\n💬 "<i>${escapeHtml((m.message_text || '').slice(0, 100))}</i>"\n📱 ${escapeHtml(m.device_type || 'Desktop')} | 🕒 <i>${m.created_at}</i>`;
    }).join('\n\n');
    const text = `📩 <b>SO'NGGI ${messages.length} TA QABUL QILINGAN XABAR:</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n${mLines || '<i>Hozircha murojaatlar yo\'q.</i>'}\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n⚡ <i>Real PostgreSQL bazasi</i>`;
    return { text, markup: getAdminSubKeyboard('admin_leads') };
  }
  if (data === 'admin_broadcast') {
    const text = `📢 <b>XABAR TARQATISH TIZIMI (BROADCAST)</b>\n━━━━━━━━━━━━━━━━━━━━━━━━━━\n👥 <b>Auditoriya:</b> <code>${botUserCount}</code> ta bot a'zolari\n\n💡 <b>Xabar yuborish:</b>\nMahalliy Python bot orqali <code>/admin</code> menyusidan <b>Xabar Tarqatish</b> tugmasini bosib barcha a'zolarga e'lon yuborishingiz mumkin.`;
    return { text, markup: getAdminSubKeyboard('admin_broadcast') };
  }
  if (data.startsWith('setlang_')) {
    return {
      text: `✅ Til muvaffaqiyatli tanlandi! Quyidagi menyudan foydalanishingiz mumkin.`,
      markup: null,
    };
  }
  return null;
}
